slackapi/deno-slack-hooks
# Add check-update hook functionality

###  Summary

This pull requests builds on the original [POC to introduce the check-update hook](https://github.com/slackapi/deno-slack-hooks/pull/1):

> This hook implementation inspects a hermes deno project to see if it uses any Slack deno SDK components, and if so, checks the versions for these SDK components in the project are up to date (by pinging deno.land).
>
> Deno is still pretty fresh and it is a bit of a wild west, so the versions of dependencies making are kind of strewn about in a few different locations (mainly due to https://github.com/denoland/deno/issues/14066). As a result, the code in this PR has to look into two files to inspect the various versions in use in the hermes project (import_map.json and .slack/slack.json). Should that aforementioned-bug ever get resolved (or perhaps we resolve it?), then we could put all of the SDK dependencies into the import_map.json

####  Updates
The updated implementation held onto the original POC's `versionMap`, but took it a step farther in that it creates a full map that features details about each dependency. Once the map is complete, it is used to create the `UpdateResponse` that the CLI is expecting, summarizing the updates available and creating an aggregate message to be displayed.

_Note: though it was tempting to forgo the `versionMap` all together, and produce just the aggregate message that the CLI expects, for extensibility purposes, I thought it might be useful to have a robust data structure that featured all findings. If, in the future, there are more complicated inter-dependencies or a specific order is desired, this sets a solid foundation from which those things can be achieved._

### ❗ Known Issues / Questions

**The formatting that's being applied is reminiscent of prettier.**
And it's made the code _significantly_ less readable, IMO, with its stacking of arguments and the like.

**Existing issue of "last wins" still exists.**
Pre-existing TODO noted that the implementation "_doesn't cover the case where multiple commands use the same SDK. Only the "last" key containing the relevant SDK will be checked, as-is_". Is this pressing enough to fix now? If so, what behavior would we like?

**Sharp edges? Dusty?**
There is some tidying up that can still be done with the code quality and I aim to chip away at that as feedback comes in.

### Testing Steps
The full experience requires the CLI to work in concert with this update. However, to test it out, in a new project in the absence of this: 

1. Run the following command (which would normally be in the `check-update` hook):

```
deno run -q --unstable --allow-read --allow-net file:///Users/<YOUR_PATH_HERE>/deno-slack-hooks/src/check-update.ts
```

2. `echo` the value returned for the `message` key of the JSON.

3. Assuming you have out-of-date dependencies, your output should look like the following (and, if not, alter your dependencies to be lower than latest):

<img width="197" alt="Screen Shot 2022-05-10 at 11 13 20 PM" src="https://user-images.githubusercontent.com/7788242/167722868-aeddfbd1-7627-4133-a0c0-a06622e7c763.png">



### Requirements (place an `x` in each `[ ]`)
